### DISCUSSION TITLE
[Proposal] Implicit lambda parameter and trailing lambda in C#

### DISCUSSION THREAD

**Discussion by ronnygunawan** [dotnet/csharplang#6122] [Language Ideas]
```csharp
var filtered = items.Where { Price > 9.99m }
                    .OrderBy { Name }
                    .ToList();
```

## Implicit lambda parameter:
- Allow omitting `() =>` when lambda has no parameter.
- Allow omitting `arg1 =>` when lambda has only one parameter.
- Lambda with implicit parameter must always be enclosed in braces.
- If lambda is ambiguous between `() =>` and `arg1 =>`, then implicit parameter can't be used.
- Members of implicit parameter can be directly used without member access operator, and will also hide names from enclosing type.
- The implicit parameter can be accessed using `value` keyword.
- `async` can be placed before the opening brace.
- `this` is unaffected. It still refers to enclosing type.

## Trailing lambda:
- If last parameter of a method is a lambda with implicit parameter, then the lambda can be written after the closing parentheses.
- If the method only have one such parameter, then the parentheses can be omitted too.
- Trailing lambda can be used in constructors as long as there is no overload that can cause conflict between trailing lambda and object initializer. eg:
```csharp
class Foo {
    public Foo(int x) { }
    public Foo(int x, Func<int, bool> f) { }
}

var foo = new Foo(10) { value % 2 == 0 }; // this is illegal because Foo has a
                                          // shorter constructor without the trailing lambda.

var foo = new Foo(10, f: { value % 2 == 0 }); // this is fine
```

## Examples:
```csharp
class Item {
    public string Name { get; set; }
    public decimal Price { get; set; }
}
```

### Functions
```csharp
// normal lambda
var filtered = items.Where(i => i.Price > 9.99m)
                    .OrderBy(i => i.Name)
                    .ToList();

// with implicit lambda parameter
var filtered = items.Where({ Price > 9.99m })
                    .OrderBy({ Name })
                    .ToList();

// with trailing lambda
var filtered = items.Where { Price > 9.99m }
                    .OrderBy { Name }
                    .ToList();
```

### Actions
```csharp
// normal lambda
items.ForEach(i => {
    Debug.WriteLine(i.Name);
    Console.WriteLine(i.Price);
});

// with implicit lambda parameter
items.ForEach({
    Debug.WriteLine(Name);
    Console.WriteLine(value.Price); // 'value' can be omitted
});

// with trailing lambda
items.ForEach {
    Debug.WriteLine(Name);
    Console.WriteLine(value.Price); // 'value' can be omitted
};
```

### Method with multiple lambda
```csharp
// normal lambda
var priceByName = items.ToDictionary(i => i.Name, i => i.Price);

// with implicit lambda parameter
var priceByName = items.ToDictionary({ Name }, { Price });

// with trailing lambda
var priceByName = items.ToDictionary({ Name }) { Price };
```

### Get creative with trailing lambda
```csharp
// VB-style With block
// note: following methods are not System.Console's static methods
With (console) {
    WriteLine("Hello World");
    BackgroundColor(ConsoleColor.Green);
    WriteLine("I am Green");
};

// fire and forget
Task.Run async {
    await Task.Delay(10000);
};

// foreach value in Range
ForEach(0..100) {
    Console.WriteLine(value);
};
```

**Comment by HaloFour** [comment]
Very similar to https://github.com/dotnet/csharplang/issues/1151

Kotlin and Scala both have a flavor of this feature (or set of features which can be combined to achieve this kind of syntax) and it can be useful to define DSL-like syntax.  Although, IMO, Scala takes it a bit too far as you can omit the dot operator (sometimes) and you can omit the parenthesis (sometimes).

**Comment by ronnygunawan** [comment]
#1151 went too far imo, it retargets what `this` refers to. In my opinion, the ability to retarget `this` doesn't fit in C#.

**Comment by orthoxerox** [comment]
@ronnygunawan wrll, your suggestion retargets `value` instead. Retargeting `this` or at least bringing the members of the parameter into scope would be very useful for writing DSLs.

**Comment by ronnygunawan** [comment]
@orthoxerox well, I think both retargeting `value` and `this` has their own problems then. Retargeting `value` limits lambda's usability when used inside a property setter, while retargeting `this` hides the original `this` from enclosing type.

Kotlin and Java solve the `this` retargeting problem using `EnclosingClassName@this`, and for the `value` retargeting problem, `value` keyword can be replaced by a new keyword `it`.

**Comment by marchy** [comment]
Is this on the consideration list for C# 11 / C# 12?

The RemObjects C# compiler supports this [Trailing Closures](https://docs.elementscompiler.com/Hydrogene/LanguageExtensions/TrailingClosures/) feature quite nicely, and Swift and Kotlin have supported this from the get-go (5+ years now). Would be great to see C# catch up to this modern must-have that can unlock all sorts of DSL-style APIs.

_NOTE: Swift has since extended to support [multiple trailing closures](https://www.hackingwithswift.com/swift/5.3/multiple-trailing-closures) (full [feature spec](https://github.com/apple/swift-evolution/blob/master/proposals/0279-multiple-trailing-closures.md)), thus directly supporting scenarios such as the example given above:_


_STRETCH: Not sure if compatible with C# named parameters overall, but being able to optionally specify those as well would make it these much more readable/useful as well._


**Some very useful real-world examples from functional-style utilities:**
``` C#
//
// EXAMPLE 1
//
// current lambdas
DateTimeOffset rangeEndDate = endDateTime?.IfNotNull( endDateTime_ =>
		endDateTime_.StartOfDay() - ((crossesMidnight) ? 1 : 0).Days(),
	otherwise: () => startDateTime.StartOfDay() );

// with trailing lambda
DateTimeOffset rangeEndDate = endDateTime?.IfNotNull(
	value.StartOfDay() - ((crossesMidnight) ? 1 : 0).Days()
){ startDateTime.StartOfDay() };

// with multiple & named trailing lambdas (STRETCH)
DateTimeOffset rangeEndDate = endDateTime?.IfNotNull {
	value.StartOfDay() - ((crossesMidnight) ? 1 : 0).Days()
}, otherwise: { startDateTime.StartOfDay() };


//
// EXAMPLE 2
//
// current lambdas
GeoCoordinate? coordinate = Try.CatchingAsNull<ArgumentOutOfRangeException, GeoCoordinate?>( () =>
	new GeoCoordinate( place.geoCoordinate.latitude, place.geoCoordinate.longitude ),
@catch: x => { // NOTE: indentation hack to make it "feel" like a trailing lambda
	Log.LogError( $"Geo-coordinate for content #{attraction.id} out of bounds: ({place.geoCoordinate.latitude}, {place.geoCoordinate.longitude}). Error: {x.Message}" );
});

// with trailing lambda
GeoCoordinate? coordinate = Try.CatchingAsNull<ArgumentOutOfRangeException, GeoCoordinate?>( 
	new GeoCoordinate( place.geoCoordinate.latitude, place.geoCoordinate.longitude )
){ // NOTE: no indentation hack. Trailing lambdas naturally style back to the containing statement's indentation
	Log.LogError( $"Geo-coordinate for content #{place.id} out of bounds: ({place.geoCoordinate.latitude}, {place.geoCoordinate.longitude}). Error: {value.Message}" );
};

// with multiple & named trailing lambdas (STRETCH)
GeoCoordinate? coordinate = Try.CatchingAsNull<ArgumentOutOfRangeException, GeoCoordinate?>{
	new GeoCoordinate( place.geoCoordinate.latitude, place.geoCoordinate.longitude )
}, @catch: x => {
	Log.LogError( $"Geo-coordinate for content #{attraction.id} out of bounds: ({place.geoCoordinate.latitude}, {place.geoCoordinate.longitude}). Error: {x.Message}" );
};
```

**Comment by D0ctorWh0** [comment]
This could be very useful to convert code in fluent style into something easier to read and write. For example this 

```csharp
FluentQuery
	.ConnectionString("connectionString")
	.StoredProcedure("LoadData")
	.Parameters(c=>c
		.AddParameter("ParName1", "ParValue1")
		.AddParameter("ParName2", "ParValue3")
		.AddParameter("ParName3", "ParValue3")
		)
	.Execute();
```

into this

```csharp
FluentQuery
	.ConnectionString("connectionString")
	.StoredProcedure("LoadData")
	.Parameters
	{
		AddParameter("ParName1", "ParValue1")
		AddParameter("ParName2", "ParValue3")
		AddParameter("ParName3", "ParValue3")
	}
	.Execute();
```

or even this

```csharp
FluentQuery
{
	ConnectionString("connectionString")
	StoredProcedure("LoadData")
	Parameters
	{
		AddParameter("ParName1", "ParValue1")
		AddParameter("ParName2", "ParValue3")
		AddParameter("ParName3", "ParValue3")
	}
	Execute()
};
```

It seems to me it looks much easier to write and read.

**Comment by pelletburner** [comment]
Would allow for much cleaner use of Parallel.ForEach and several AOP-like patterns, too. I support this.

**Comment by tmteam** [comment]
I write a lot of code in Kotlin, where this syntax exists - and this is a game changer.

1) it allows you to move away from the fluent api towards the DSL api. In Compare to Kotlin - it is HUGE benefit of Kotlin

2) it is critical for the functional style of writing code. Imagine 3 or 4 nested lambdas, ugly mess - that is my everyday live

3) it is critical for Structural concurrency tasks. Here, braces are boundaries to concurrency scope, so it keeps code clean

**Comment by jantajov** [comment]
So, is this conaidered for upcoming c# releases?

**Reply by CyrusNajmabadi** [reply]
@jantajov it is not.

**Reply by iam3yal** [reply]
@jantajov I doubt it or at least this specific issue isn't championed yet.

**Reply by marchy** [reply]
Most impactful C# feature that could be taken on by a long-shot.

Multiple languages already implemented it so abundance of reference points to draw upon.

Yet still isn’t championed.

Super sad 😔

**Reply by CyrusNajmabadi** [reply]
There's been very little community interest on this issue versus many others. I don't think it's necessarily as impactful to the ecosystem as it would be to your personal work :-)

**Reply by JanTajovKronesDigital** [reply]
@CyrusNajmabadi How can we push this feature? I bet a lot of people doesn't even know this discussion exists, no wonder it doesn't spark lot of interest.

**Reply by jstafford5380** [reply]
> There's been very little community interest on this issue versus many others. I don't think it's necessarily as impactful to the ecosystem as it would be to your personal work :-)

I think it would be. So much of c# is fluent sytnax now anyway, this would clean it up. Also opens some doors for coded UIs since the model can be closer to markup. Less gear shifting. Could help MAUI quite a bit.

**Reply by marchy** [reply]
> There's been very little community interest on this issue versus many others. I don't think it's necessarily as impactful to the ecosystem as it would be to your personal work :-)

Sorry @CyrusNajmabadi  I think you are deeply misguided/uninformed about the state of modern languages.

This has precisely zero to do with "personal work". Do you have ANY idea how SwiftUI and Kotlin's Compose work or what impact they have made on being able to evolve specifying nested information outside of XML-constrained "markup" (ie: XAML 🤮)?

**Compose**
``` Kotlin
Button(
    onClick = {
        logger.info( "Pressed!" )
    }
){
    Text( "Hello $platform" )
})
```

**Swift UI**
``` Swift
struct AlbumDetail: View {
    var album: Album

    var body: some View {
        List( album.songs ){ song in 
            HStack {
                Image( album.cover )
                VStack( alignment: .leading ){
                    Text( song.title )
                    Text( song.artist.name )
                        .foregroundStyle( .secondary )
                }
            }
        }
    }
}
```

They both use trailing lambdas in order to enable a whole world of DSL's that can use **nesting** via braces `A { B { .. }}` rather than having to resort to non-nestable (and deeply over-verbose) method call parentheses `A(() => { B(() => { ... })})`.

You know – the exact same benefit that drove C#'s innovative `await` / `async` keywords over nested callbacks way back in the day?

This is THE most fundamental holdback of the entire C# ecosystem preventing both micro-usage libraries (such as I described above) as well as era-sized frameworks (such as Compose / SwiftUI).

Your dismissal of its significance only shows how ignorant you are of anything outsize of the limited .NET world you are constrained within.


When you are ready to move on from focusing on useless syntax sugar such as Primary Constructor (worst feedback/reception of any feature likely ever released in C#) and create real **transformative** value in the language – go inform yourself on this topic and please get back to bringing C# into the modern era. It has now been 10 years since both Swift and Kotlin released with this [major feature](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/closures/#Trailing-Closures) in place. You don't need to innovate. Just... catch up and stop acting as if you know better than people with broader sets of experiences than you who are trying to give you genuine feedback on the team and language's blindspots.

**Reply by CyrusNajmabadi** [reply]
> I think it would be

That's what I'm saying. We have to gauge benefit to the community over individual interest. 

> Could help MAUI quite a bit.

There's been no interest conveyed to us from Maui that this is something they're lacking or would adopt if present. If that changed from then, this would affect the assessment.

**Reply by CyrusNajmabadi** [reply]
> Do you have ANY idea how SwiftUI and Kotlin's Compose work

I'm very familiar yes. 

> and stop acting as if you know better

As I said, the community has itself shown little interest here. We're investing efforts in the areas the community, ecosystem, and partners have been asking for. That may change here in the future, but currently is where we stand.

**Reply by 333fred** [reply]
@marchy personal attacks are not acceptable here, so I have hidden your comment. Fwiw, several members of the LDM _do_ have experience with kotlin and/or swift, including using their UI DSLs. While I agree that they are pleasant to use, adopting them is not "just" a matter of adding trailing lambdas to the language. The paradigm they use is very different to our desktop frameworks, and is actually closest to how Razor works (in terms of layout updates and redrawing). Changing our desktop frameworks to support MVU in a performant manner is much bigger than just adopting this feature, and isn't even blocked by this feature: for example, the [Comet](https://github.com/dotnet/Comet) proof-of-concept used object initializers, rather than trailing lambdas, to form the core of the dsl. That does have some disadvantages, but further exploring that area, quantifying those disadvantages, and showing how the existing C# features in this area don't stand up to the use case is the best path forward.

Personally, while I quite like the idea of this feature, I'm not sure that it passes the 100 points line unless it has some attached framework interest in consumption. I haven't yet seen that, so I haven't yet championed it.

**Reply by CyrusNajmabadi** [reply]
Fwiw, I would expect first to see a prototype of such a ui framework showing how it would want to be built with lambdas/delegates, and that there was a serious view that such a path would be desirable and viable. If that existed, the idea of a trailing lambda arg (and potentially other lambda improvements) could be more desirable. 

But it's certainly not something that is going to see interest without firm belief it would actually be used and supplants the current patterns.

**Reply by marchy** [reply]
@333fred I'm all for keeping the community constructive – my attack was no more personal than @CyrusNajmabadi's snark's dismissal of my [thoughtful recommendations](https://github.com/dotnet/csharplang/discussions/6122#discussioncomment-2747395) as "personal work".

Please kindly either hide his comment or unhide mine.
You are either listening to the community or you aren't.

I made sure to refine my message as to keep it constructive as informative. I explicitly called out the extra capability this provides – modelling hierarchical information naturally in the language VS the non-nested callback-style methods that currently hold it back. Your colleague neither acknowledged this crux nor the community frameworks it holds back (I've previously brought up example from testing frameworks ie: AutoMapper and .NET's own existing frameworks LINQ etc.).

Please uncensor my comment and if you have a valid argument – stand for it.

Your censorship of real feedback will only get you NO feedback from the people who care most about the ecosystem.

**Reply by CyrusNajmabadi** [reply]
@marchy that was in regards to this being driven by personal anecdotes vs a large set of feedback indicating widespread desire from the community.  

We listen closely to feedback, but that also means we see which areas have high interest and which do not. 

As an example, while you may not like primary constructors, or believe they were worthwhile, that doesn't jive with the interest and feedback we got on those. 

That's ok though. Any decisions we make will certainly not align with all our users. We need to listen, engage, and make judgement calls, knowing that some will disagree and wish our efforts were spent elsewhere. 

If it helps, I've proposed many a language feature myself, only to see little interest out value seen by others. That's ok. My goal at that point is to then see what I can do to make things more palatable for a larger group (which often hasn't been possible). That's the fun and challenge of it all.

**Reply by 333fred** [reply]
It doesn't appear to me that Cyrus was specifically responding to the comment you linked; if he was, he would have responded there. His comment is also not a personal attack, as calling people "ignorant" is. That is a clear violation of our code of conduct. Again, please refrain from personal attacks.

**Reply by CyrusNajmabadi** [reply]
@marchy we have a code of conduct here that we ask people to abide by. Both in terms of people and posts, easily 99.99% go through (probably higher). Please just keep things civil and on topic, and please just remember that disagreement and differing opinions are a good thing. It's through that process that we come up with better designs, and more long lasting ideas that fit well with a language that is already 25 years old, and wants to keep going at least double that :-)

Thanks!
